data_processors/pipeline/lambdas/bcl_convert.py

```diff
@@ -193,6 +198,66 @@ def get_settings_by_instrument_type_assay(instrument, sample_type, assay):
     return {}
 
 
+# TODO: extract into utils?
+def get_library_id_from_sample_name(sample_name: str):
+    # format: samplename_libraryid_extension
+    # we are only interested in the library ID
+    fragments = sample_name.split("_")
+    # if there is an extension, the library ID is the second to last fragment
+    if "_topup" in sample_name or "_rerun" in sample_name:
+        return fragments[-2]
+    # if not, then the library ID is the last fragment
+    return fragments[-1]
+
+
+# TODO: extract SampleSheet handling/utils into separate module?
+def get_sample_names_from_samplesheet(gds_volume: str, samplesheet_path: str) -> List[str]:
+    if not samplesheet_path.startswith(os.path.sep):
+        samplesheet_path = os.path.sep + samplesheet_path
+    logger.info(f"Extracting sample names from gds://{gds_volume}{samplesheet_path}")
+
+    ntf: NamedTemporaryFile = gds.download_gds_file(gds_volume, samplesheet_path)
+    if ntf is None:
+        reason = f"Abort extracting metadata process. " \
+                 f"Can not download sample sheet from GDS: gds://{gds_volume}{samplesheet_path}"
+        logger.error(reason)
+        raise ValueError(reason)
+
+    logger.info(f"Local sample sheet path: {ntf.name}")
+    sample_names = set()
+    with closing(ntf) as f:
+        samplesheet = SampleSheet(f.name)
+        for sample in samplesheet:
+            sample_names.add(get_library_id_from_sample_name(sample.Sample_ID))
+
+    logger.info(f"Extracted sample names: {sample_names}")
+
+    return list(sample_names)
+
+
+def get_metadata_df(gds_volume_name, sample_sheet_gds_path) -> pd.DataFrame:
+    # Get libraries and metadata associated with this run/SampleSheet
+    sample_names: List[str] = get_sample_names_from_samplesheet(gds_volume=gds_volume_name,
+                                                                samplesheet_path=os.path.sep + sample_sheet_gds_path)
+
+    metadata_df: pd.DataFrame = pd.DataFrame()
+    for sample_name in sample_names:
+        lib_id = get_library_id_from_sample_name(sample_name)
```

Comment:
Pure `library_id` is not unique at the mo, hence, metadata query by this `library_id` might not be precise. And, we can improve this with query like `get_metadata_by_sample_library_name(...)`, here, `sample_library_name` == `Sample_ID` as in SampleSheet by using db built-in Concat function.

Reply:
The `library_id` should be unique and I think is flagged as such in the data model.
I also want to move away from the `sample_name` as quickly as possible, as it is not well defined (at least not for older data). Ideally the SampleSheet will follow suite... 
I can't see any issues using `library_id`? Did I miss something?

Reply:
At this point, 
- `sample_names` contain list of `Sample_ID` as in from SampleSheet. 
- And, `get_library_id_from_sample_name(sample_name)` will strip Library ID's `_topup` and `_rerun` extensions. 
- And, in database LabMetadata table, `library_id` is unique, only with `_topup` and `_rerun` extensions. 

So, query using stripped-pure Library ID (without extensions) may return different record that has matched, typically, older entry.

May be, let do a quick Zoom chat on this?

Reply:
Yup, that should not be an issue. If it is then it's wrong and we'd have to fix the data.
Happy for a quick Zoom chat.

Reply:
As discussed, to use Library ID on moving forward and, rolled back query with Library ID in https://github.com/umccr/data-portal-apis/commit/81298f524cce1f18acd61eaba749c3215d6a63aa